### DISCUSSION TITLE
Successfully compiling invalid razor

### DISCUSSION THREAD

**Discussion by gt5700** [toddams/RazorLight#486] [Q&A] [answered]
Hi,
 I have a razor page, being compiled from a string, that has a reference to a  property in the Model object that does not exist. My expectation was the razor page should not compile. I use this approach to allow my users to enter razor templates into a text box, and hit a "Validate" button to confirm the razor is valid before it can be saved in the database.

A simple example is as follows. The Model has a property called "Items" which is a simple List\<string\>.

```
@if(Model.Items.Count > 0)
{
    <div>Items</div>
    @foreach(var item in Model.ItemsTypo)
    {
        <li>@item</li>
    }
}
```
 You can see I have deliberately misspelt the property name in the foreach statement. This razor compiles successfully, and when rendered, it results in empty output.

 What am I missing here?

**Comment by jzabroski** [comment, marked as answer]
I don't see a `@model Type` expression here, like this cshtml file in the tests folder hierarchy:

https://github.com/toddams/RazorLight/blob/master/tests/RazorLight.Tests/Assets/Files/Partial.cshtml

If they're saving template files to the database without typed models, then I guess your validation is a bit simple.

It would be pretty cool if you shared the full example and contributed it back to the project, as others could probably benefit and improve upon it. For example, ideally you'd use OmniSharp https://github.com/OmniSharp/omnisharp-roslyn and BlazorWASM client-side to get deep intellisense support.  It might slow down typing responsiveness, though.

But, it would be pretty powerful.  You can imagine dragging and dropping from an AST web control elements from the Model into a ProseMirror or Monaco Editor environment.  For ProseMirror, most users wouldnt even need to know its Razor.  It would just look like a fancy mail merge from MS Office.

**Reply by gt5700** [reply]
Thanks @jzabroski , that's pointed me in the correct direction. It seems the Razor library I'm migrating from (RazorEngine) must have been setting the `@model <Type>` behind the scenes. I've added in code now above the code the user puts into the text box, and it causes the compilation to fail as expected. Awesome!

What I've learnt now is that if the model type is not set, then it appears to be a "dynamic" type.

I also learnt that I was so focussed on why the the compilation was successful, that I failed to see the obvious reason why it didn't fail at runtime. I wasn't forcing a "typo" in the `@if` statement, and my test data had an empty list, so the foreach statement was never reached, to trigger a runtime error for the invalid property name!

My current solution is very basic. The razor the user enters into the web page text box can be used at runtime with multiple "similar" models. This is why I don't hardwire the exact model in what is stored in the database. When they click the "Validate" button, I collect all possible models the Razor could be "run against", and try to compile (now with the `@model <Type>` set at the top).

I will definitely look at your suggestions above, as they sound quite powerful. My software has a lot of places where I have text input and I'm using ACE editor for most of them. It would be great if I could give deeper "Intellisense" capability. ACE editor is great, as it does support simple "intellisense" features by scanning the current contents of the text, and offering those, but true "Intellisense" would be awesome.

**Reply by jzabroski** [reply]
I'm familiar with ACE Editor.  I was briefly the maintainer of CLEditor before Guardian unveiled Scribe.  Subsequently, ProseMirror, ACE Editor and Monaco all came out.

It would be great if you could contribute back.  We would love a third maintainer on this project.  Ivan is currently serving in war, and I'm about to have a kid.  I'm interested in keeping this project alive, but time constraints have me quite a bit.  I also run FluentMigrator for db migrations.
